crates/sui-storage/src/mutex_table.rs

```diff
@@ -10,24 +10,75 @@ use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
 use std::sync::Arc;
 use std::time::Duration;
 
-use tokio::sync::{Mutex, RwLock};
+use async_trait::async_trait;
+use tokio::sync::{
+    Mutex, OwnedMutexGuard, OwnedRwLockReadGuard, OwnedRwLockWriteGuard, RwLock, TryLockError,
```

Comment:
@andll Can you have a look at the locks use here? You mentioned you wanted to examine the use of `tokio::sync` lock primitives.

Reply:
I think we should avoid using tokio::locks in favor of parking_lot locks where possible.
Most of the times it is possible, unless there is a place where we need to hold a lock across await points.

Looking at the code below, it seems that we hold lock across await point on the `update_objects_and_locks` function, but it also seems that this function actually does not need to be async. So if possible, I would replace tokio locks with parking lot locks in this PR.

(And indeed if we could just avoid adding new locks that would be great, but I am assuming this was considered and we did not figure out any alternative)

Reply:
this PR employs existing API (i.e. MutexTable), just slightly extends it to RwLocks.
It feels like migration to parking_lot locks for MutexTable should be done in separate PR. 
It requires review and conversion of existing call sites, FWIK there are two more.
But yes, new call site is not a blocker for conversion and doesn't have to be async

Reply:
FWIW, I am not suggesting rewriting existing locks, I am just suggesting not adding new async locks :)

Reply:
looks like other call site can be also converted to sync lock: https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority/authority_store.rs#L809
not sure about WAL use case though: https://github.com/MystenLabs/sui/blob/main/crates/sui-storage/src/write_ahead_log.rs#L337

cause if they both can be converted to sync, I'd prefer to do that in follow up PR, instead of creating copy/paste implementation for synchronous RwLockTable just for this new case

Reply:
MutexTable cannot be converted to parking_lot wholesale, as the WAL case requires being held across awaits.

I think it shouldn't be too hard to have a sync version of MutexTable, although some code duplication may be needed since its hard to be generic over async-ness.

Reply:
@mystenmark I am curious why WAL locks need to be held across await points, could you point me code?

I briefly looked and it seemed that we use lock within same task, and we just happen to have bunch of async functions like `update_objects_and_locks` that actually don't need to be async. But maybe I missed something, since I did not look too much

Reply:
per discussion with @andll, let's refactor the code and convert to sync rwlock in the followup PR. We'll also convert `acquire_transaction_locks` use case